File: app/src/main/java/lee/jaebeom/allwiki/main/MainContentFragment.java
package lee.jaebeom.allwiki.main;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.JavascriptInterface;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.ProgressBar;

import java.io.Serializable;

import lee.jaebeom.allwiki.R;
import lee.jaebeom.allwiki.util.DOMAccess;
import lee.jaebeom.allwiki.util.NestedWebView;
import lee.jaebeom.allwiki.util.SharedPreference;

/**
 * Created by leejaebeom on 2017. 10. 14..
 */

public class MainContentFragment extends Fragment implements MainContract.TabView, View.OnKeyListener, Serializable {
    String URL;
    String wiki;
    MainContract.TabPresenter presenter;
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null){
            wiki = getArguments().getString("wiki");
        }
    }

    public static MainContentFragment newInstance(String wiki) {
        Bundle args = new Bundle();
        args.putString("wiki", wiki);
        MainContentFragment fragment = new MainContentFragment();
        fragment.setArguments(args);
        return fragment;
    }
    NestedWebView webView;
    WebView DOMAccessWedView;
    ProgressBar webLoadingBar;
    @SuppressLint({"SetJavaScriptEnabled", "JavascriptInterface"})
    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (savedInstanceState == null){
            if (!SharedPreference.getStringPreferences(getContext(),SharedPreference.keyKeyword).equals("")){
                URL = presenter.getWikiURL(getArguments().getString("wiki")) + SharedPreference.getStringPreferences(getContext(),SharedPreference.keyKeyword);
            }else{
                URL = presenter.getWikiFrontURL(getArguments().getString("wiki"));
            }
        }else{
            URL = savedInstanceState.getString("URL");
            wiki = savedInstanceState.getString("wiki");
        }

        View view = inflater.inflate(R.layout.fragment_main, container, false);;
        webView = view.findViewById(R.id.fragment_web_view);
        DOMAccessWedView = (WebView)view.findViewById(R.id.fragment_DOM_Access);
        webLoadingBar = view.findViewById(R.id.fragment_progress_bar);

        DOMAccessWedView.getSettings().setJavaScriptEnabled(true);
        DOMAccessWedView.addJavascriptInterface(new MyJavaScriptInterface(), "HTMLOUT");

        webView.getSettings().setJavaScriptEnabled(true);
        webView.canScrollHor(10);

        DOMAccessWedView.setWebViewClient(new WebViewClient(){
            @Override
            public void onPageStarted(WebView view, String url, Bitmap favicon) {
                webLoadingBar.setVisibility(View.VISIBLE);
            }

            @Override
            public void onPageFinished(WebView view, String url) {
                view.loadUrl(DOMAccess.getJavascript(wiki));

            }
        });

        DOMAccessWedView.setWebChromeClient(new WebChromeClient(){
            @Override
            public void onProgressChanged(WebView view, final int newProgress) {
                webLoadingBar.setProgress(newProgress);
            }
        });
        webView.setWebViewClient(new WebViewClient(){

            @TargetApi(Build.VERSION_CODES.N)
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
                DOMAccessWedView.loadUrl(request.getUrl().toString());
                return true;
            }

            @SuppressWarnings("deprecation")
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                DOMAccessWedView.loadUrl(url);
                return true;
            }

            @Override
            public void onPageFinished(WebView view, String url) {
                webLoadingBar.setVisibility(View.GONE);
            }
        });

        webView.setOnKeyListener(this);

        DOMAccessWedView.loadUrl(URL);
        return view;
    }

    @Override
    public void setPresenter(MainContract.TabPresenter presenter) {
        this.presenter = presenter;
    }

    @Override
    public void CallBack() {
        webView.post(new Runnable() {
            @SuppressLint("SetJavaScriptEnabled")
            @Override
            public void run() {
                webView.loadDataWithBaseURL(URL, presenter.getHTML(), "text/html", "UTF-8", null);

            }
        });

    }

    @Override
    public boolean onKey(View view, int i, KeyEvent keyEvent) {
        if (i == KeyEvent.KEYCODE_BACK && keyEvent.getAction() == MotionEvent.ACTION_UP && webView.canGoBack()){
            webView.goBack();
            webLoadingBar.setVisibility(View.VISIBLE);
            return true;
        }
        return false;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putString("URL", URL);
        outState.putString("wiki", wiki);
        super.onSaveInstanceState(outState);
    }

    class MyJavaScriptInterface {

        @JavascriptInterface
        public void processHTML(final String html) {
            presenter.saveHTML(html);
            presenter.CallBack();
        }
    }
}

File: app/src/main/java/lee/jaebeom/WikiPortal/main/MainContentFragment.kt
package lee.jaebeom.WikiPortal.main

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.support.v4.app.Fragment

import android.databinding.DataBindingUtil
import android.graphics.Bitmap
import android.os.Build
import android.os.Bundle
import android.view.*
import android.webkit.*
import lee.jaebeom.WikiPortal.R
import lee.jaebeom.WikiPortal.databinding.FragmentMainBinding
import lee.jaebeom.WikiPortal.main.MainContract.TabPresenter
import lee.jaebeom.WikiPortal.util.DOMAccess
import lee.jaebeom.WikiPortal.util.SharedPreference
import java.io.Serializable

/**
 * Created by leejaebeom on 2017. 10. 31..
 */
class MainContentFragment : Fragment(), MainContract.TabView, View.OnKeyListener, Serializable {
    lateinit var presenter: TabPresenter
    lateinit var wiki: String
    lateinit var URL: String
    lateinit var binding :FragmentMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        wiki = arguments.getString("wiki")
    }

    companion object {
        @JvmStatic fun newInstance(wiki: String) : MainContentFragment{
            val args = Bundle()
            args.putString("wiki", wiki)
            val fragment = MainContentFragment()
            fragment.arguments = args
            return fragment
        }
    }

    @SuppressLint("SetJavaScriptEnabled", "SetJavaScriptEnabled")
    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View {
        if (presenter != null){
            if (!SharedPreference.getStringPreferences(activity.applicationContext, SharedPreference.keyKeyword).equals("")){
                URL = presenter.getWikiURL(arguments.getString("wiki")) + SharedPreference.getStringPreferences(activity.applicationContext, SharedPreference.keyKeyword)
            }else{
                URL = presenter.getWikiFrontURL(arguments.getString("wiki"))
            }
        }else{
            URL = savedInstanceState?.getString("URL")!!
            wiki = savedInstanceState.getString("wiki")!!
        }
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_main, container, false)
        val view = binding.root

        binding.fragmentDOMAccess.settings.javaScriptEnabled = true
        binding.fragmentDOMAccess.addJavascriptInterface(MyJavaScriptInterface(), "HTMLOUT")

        binding.fragmentWebView.settings.javaScriptEnabled = true
//        binding.fragmentWebView.canScrollHor(10)

        binding.fragmentDOMAccess.webViewClient = object : WebViewClient(){
            override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
                binding.fragmentProgressBar.visibility = View.VISIBLE
            }

            override fun onPageFinished(view: WebView?, url: String?) {
                view?.loadUrl(DOMAccess.selectJavascript(wiki))
            }
        }

        binding.fragmentDOMAccess.webChromeClient = object : WebChromeClient(){
            override fun onProgressChanged(view: WebView?, newProgress: Int) {
                binding.fragmentProgressBar.progress = newProgress
            }
        }

        binding.fragmentWebView.webViewClient = object : WebViewClient(){
            @TargetApi(Build.VERSION_CODES.N)
            override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
                binding.fragmentDOMAccess.loadUrl(request?.url.toString())
                return true
            }

            @Suppress("DEPRECATION")
            override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean {
                binding.fragmentDOMAccess.loadUrl(url)
                return true
            }

            override fun onPageFinished(view: WebView?, url: String?) {
                binding.fragmentProgressBar.visibility = View.GONE
            }
        }

        binding.fragmentWebView.setOnKeyListener(this)

        binding.fragmentDOMAccess.loadUrl(URL)
        return view
    }

    override fun _presenter(presenter: TabPresenter) {
        this.presenter = presenter
    }

    override fun CallBack() {
        binding.fragmentWebView.post(Runnable {
            binding.fragmentWebView.loadDataWithBaseURL(URL, presenter.getHTML(), "text/html", "UTF-8", null)
        })
    }

    override fun onKey(p0: View?, i: Int, keyEvent: KeyEvent): Boolean {
        if (i == KeyEvent.KEYCODE_BACK && keyEvent.action == MotionEvent.ACTION_UP && binding.fragmentWebView.canGoBack()){
            binding.fragmentWebView.goBack()
            binding.fragmentProgressBar.visibility = View.VISIBLE
            return true
        }
        return false
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        outState?.putString("URL", URL)
        outState?.putString("wiki", wiki)
        super.onSaveInstanceState(outState)
    }

    inner class MyJavaScriptInterface{
        @JavascriptInterface
        fun processHTML(html: String) {
            presenter.saveHTML(html)
            presenter.CallBack()
        }
    }

}